File: backend/app/api/v1/endpoints/public_notifications.py
"""Public notification endpoints — accessible by session_id without authentication."""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query

from app.api.v1.dependencies import get_notification_service
from app.schemas.notification import NotificationResponse
from app.schemas.page import Page
from app.services.notification import NotificationService

router = APIRouter()


@router.get("/public", response_model=Page[NotificationResponse])
async def list_public_notifications(
    session_id: Annotated[str, Query(min_length=1, max_length=255)],
    notification_service: Annotated[NotificationService, Depends(get_notification_service)],
    offset: Annotated[int, Query(ge=0)] = 0,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
):
    items, total = await notification_service.get_public_notifications(session_id, offset, limit)
    return Page[NotificationResponse](
        items=[NotificationResponse.model_validate(item) for item in items],
        total=total,
        offset=offset,
        limit=limit,
    )


@router.post("/{notification_id}/read", status_code=204)
async def mark_notification_as_read(
    notification_id: UUID,
    session_id: Annotated[str, Query(min_length=1, max_length=255)],
    notification_service: Annotated[NotificationService, Depends(get_notification_service)],
):
    await notification_service.mark_as_read(notification_id, session_id)


@router.delete("/{notification_id}", status_code=204)
async def delete_public_notification(
    notification_id: UUID,
    session_id: Annotated[str, Query(min_length=1, max_length=255)],
    notification_service: Annotated[NotificationService, Depends(get_notification_service)],
):
    await notification_service.delete_notification(notification_id, session_id)

File: frontend/src/lib/api/public-notifications.ts
import { apiClient } from "@/lib/api/client"
import type { Page } from "@/types/api"
import type { Notification } from "@/types/notifications"

function buildQuery(params: Record<string, string | number>): string {
  const sp = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null && v !== "") {
      sp.set(k, String(v))
    }
  }
  const qs = sp.toString()
  return qs ? `?${qs}` : ""
}

export const publicNotificationsApi = {
  list: (sessionId: string, offset = 0, limit = 20) =>
    apiClient.get<Page<Notification>>(
      `/notifications/public${buildQuery({ session_id: sessionId, offset, limit })}`
    ),

  markAsRead: (notificationId: string, sessionId: string) =>
    apiClient.post<void>(
      `/notifications/${notificationId}/read${buildQuery({ session_id: sessionId })}`
    ),

  delete: (notificationId: string, sessionId: string) =>
    apiClient.delete<void>(
      `/notifications/${notificationId}${buildQuery({ session_id: sessionId })}`
    ),
}
